### PR TITLE
Another round of fixes to pass many more unit tests

### PR DESCRIPTION
**Pull request by haberman**
Newly-implemented functionality:
* Implemented `msg.IsInitialized()`
* Implemented `msg.FOO_FIELD_NUMBER` constants
* Implemented `repeated_field.reverse()`
* Implemented iteration over the `msg.Extensions` dict.

Fixes/refinements:
* If a string field contains invalid UTF-8, return it as a bytes object instead.
* Allow bytes attributes in `getattr(msg, attr)`
* Prohibit implicit float->int conversions.
* Refine the exception type returned when attempting an invalid assignment to an extension field.
* Make `msg.Clear()` properly detach any non-reified fields.
* If `msg.SerializeToString()` fails due to unset required fields, put the names of those fields in the error message.
* Sort the list of fields returned from `msg.ListFields()` by field number.
* Assigning to a slice now works properly if the slice is a different length than the sequence being assigned.
* Text format for map entries now omits the colon, to match C++.

### CODE REVIEW

File: python/convert.c [hunk 1]
```diff
@@ -51,8 +51,19 @@ PyObject* PyUpb_UpbToPy(upb_msgval val, const upb_fielddef *f, PyObject *arena)
...
+    case UPB_TYPE_STRING: {
+      PyObject* ret =
+          PyUnicode_DecodeUTF8(val.str_val.data, val.str_val.size, NULL);
+      // If the string can't be decoded in UTF-8, just return a bytes object
+      // that contains the raw bytes. This can't happen if the value was
+      // assigned using the members of the Python message object, but can happen
+      // if the values were parsed from the wire (binary).
+      if (ret == NULL) {
+        PyErr_Clear();
+        ret = PyBytes_FromStringAndSize(val.str_val.data, val.str_val.size);
```
[the remark this answers is not in the document]
Comment (haberman) [reply python/convert.c:63]: Yes it matches existing behavior. I agree it's weird.

File: python/convert.c [hunk 2]
```diff
@@ -65,9 +76,13 @@ PyObject* PyUpb_UpbToPy(upb_msgval val, const upb_fielddef *f, PyObject *arena)
 }
 
 static bool PyUpb_GetInt64(PyObject *obj, int64_t *val) {
+  // We require that the value is either an integer or has an __index__
+  // conversion.
+  if (!PyIndex_Check(obj)) {
+    PyErr_Format(PyExc_TypeError, "Expected integer: %S", obj);
```
[the remark this answers is not in the document]
Comment (haberman) [reply python/convert.c:82]: Done.

File: python/convert.c [hunk 3]
```diff
@@ -80,17 +95,21 @@ static bool PyUpb_GetInt64(PyObject *obj, int64_t *val) {
...
 
 static bool PyUpb_GetUint64(PyObject *obj, uint64_t *val) {
-  // For uint64 Python does not offer any functions as convenient as
-  // PyLong_AsLongLong().  If the object is not already a "long" we must
-  // manually perform the automatic conversion (using __index__() or __int__())
-  // that users expect.
+  // We require that the value is either an integer or has an __index__
+  // conversion.
+  if (!PyIndex_Check(obj)) {
+    PyErr_Format(PyExc_TypeError, "Expected integer: %S", obj);
```
[the remark this answers is not in the document]
Comment (haberman) [reply python/convert.c:101]: Done.

File: python/message.c [hunk 4]
```diff
@@ -919,6 +928,71 @@ static PyObject* PyUpb_CMessage_HasField(PyObject* _self, PyObject* arg) {
...
+  return PyBool_FromLong(ret);
+}
+
+static PyObject* PyUpb_CMessage_IsInitialized(PyObject* _self, PyObject* args) {
+  PyObject* errors = NULL;
+  if (!PyArg_ParseTuple(args, "|O", &errors)) {
+    return NULL;
+  }
+  upb_msg* msg = PyUpb_CMessage_GetIfReified(_self);
+  if (!msg) Py_RETURN_NONE;  // TODO
```
[the remark this answers is not in the document]
Comment (haberman) [reply python/message.c:950]: Fixed it instead.

File: python/message.c [hunk 5]
```diff
@@ -919,6 +928,71 @@ static PyObject* PyUpb_CMessage_HasField(PyObject* _self, PyObject* arg) {
...
+                                                  PyObject* val) {
+  assert(PyTuple_Check(val));
+  PyObject* field = PyTuple_GetItem(val, 0);
+  const upb_fielddef* f = PyUpb_FieldDescriptor_GetDef(field);
+  return PyLong_FromLong(upb_fielddef_number(f));
+}
+
+static bool PyUpb_CMessage_SortFieldList(PyObject* list) {
+  PyUpb_ModuleState* state = PyUpb_ModuleState_Get();
+  bool ok = false;
```
[the remark this answers is not in the document]
Comment (haberman) [reply python/message.c:975]: Done.

File: python/message.c [hunk 6]
```diff
@@ -919,6 +928,71 @@ static PyObject* PyUpb_CMessage_HasField(PyObject* _self, PyObject* arg) {
...
+  ret = PyObject_Call(method, args, kwargs);
+  if (!ret) goto err;
+  ok = true;
+
+ err:
+  Py_XDECREF(method);
+  Py_XDECREF(args);
+  Py_XDECREF(kwargs);
+  Py_XDECREF(ret);
+  return ret;
```
[the remark this answers is not in the document]
Comment (haberman) [reply python/message.c:993]: Done. This is unfortunately one of those cases where the failures only occur with allocation failure, which is near-impossible to inject.

File: python/message.c [hunk 7]
```diff
@@ -1539,11 +1674,27 @@ static PyObject* PyUpb_MessageMeta_GetDynamicAttr(PyObject* self,
...
 
-  // TODO(haberman): *_FIELD_NUMBER attributes?  Haven't seen any failing
-  // tests for these yet.
-
   Py_DECREF(py_key);
 
+  const char* suffix =  "_FIELD_NUMBER";
+  size_t n = strlen(name_buf);
+  size_t suffix_n = strlen(suffix);
+  if (n > strlen(suffix) &&
```
[the remark this answers is not in the document]
Comment (haberman) [reply python/message.c:1682]: Done.

**Comment by haberman** [comment]
PTAL.

File: python/message.c [hunk 8]
```diff
@@ -935,10 +935,19 @@ static PyObject* PyUpb_CMessage_IsInitializedAppendErrors(PyObject* _self,
                                                           PyObject* errors) {
   PyObject* list = PyUpb_CMessage_FindInitializationErrors(_self, NULL);
   if (!list) return NULL;
-  bool ret = PyList_Size(list) == 0;
-  if (!ret) PyObject_CallMethod(errors, "extend", "O", list);
+  bool ok = PyList_Size(list) == 0;
+  PyObject* ret = NULL;
+  PyObject* result = NULL;
+  if (!ok) {
+    result = PyObject_CallMethod(errors, "extend", "O", list);
```
Comment (fowles) [inline python/message.c:942]: having objects named both `ret` and `result` is a bit confusing.  Please pick a more clear name for `result`

**Review by fowles — APPROVED** [review]

File: python/message.c (hunk 8 above)
Comment (haberman) [reply python/message.c]: Renamed to `extend_result`.

**Pull request merged**